### DISCUSSION TITLE
Updating existing containers?

### DISCUSSION THREAD

**Discussion by mcdent** [containers/podman#18885] [Q&A] [answered]
Hi
I have a series of rootless containers running from systemd files. I'd like the ability to have some of them periodically update.
As I understand from https://docs.podman.io/en/latest/markdown/podman-auto-update.1.html I can add 
`--label io.containers.autoupdate=registry` to the ExecStart line of the unit file?

Then I need to have `sudo systemctl enable podman-auto-update.timer` running to periodically update them via `podman-auto-update.service`

My main question, assuming the above is correct; can I retrospectively add the `--label io.containers.autoupdate=registry` to each of the unit files I want updating? Or does this need setting on the command line when I first create the container?

Thanks!

**Comment by vrothberg** [comment]
Thanks for reaching out, @mcdent!

Adding the generated unit files is dangerous and not something we can support.  I assume you used `podman generate systemd`. Have you heard of [Quadlet](https://www.redhat.com/sysadmin/quadlet-podman)?  That's a new way of running Podman containers under systemd and _much_ easier to use.

**Reply by mcdent** [reply]
Thanks for your reply. I'm not sure I follow. Are you saying using `podman generate systemd` to create a systemd unit file is dangerous?
Or do you mean editing these startup files, once created is dangerous?
I kind of thought this was the correct way to run my rootless containers?

And no, I have not looked at Quadlet but I shall do so.
Thanks again!

**Reply by vrothberg** [reply, marked as answer]
Apologies for being vague.

What I meant is that editing the output of `podman generate systemd` is generally not something we can support. Editing the command-line of podman as you suggested poses no threat though.

> I kind of thought this was the correct way to run my rootless containers?

It is absolutely correct to run your containers with `podman generate systemd`.  But as you might have experienced yourself, the user experience of `podman generate systemd` is not ideal.  Changing something requires to recreate the container, and then do everything again.  [Quadlet](https://www.redhat.com/sysadmin/quadlet-podman) solves this problem in a very elegant way, such that adding the auto-update label is easy and straight-forward.

Feel free to reach out if you have questions on Quadlet!

**Reply by mcdent** [reply]
Thank you that now makes complete sense 👍  I shall go and check Quadlet out, rather than spending time figuring out the creation of all my containers again :)

**Reply by vrothberg** [reply]
Excellent!  Thanks again for reaching out and happy hacking :+1:

**Reply by mcdent** [reply]
Oooh, one of my next projects was to look at converting some of my containers over to k8s. It looks like starting with Quadlet might also help with this process too, if I am reading correctly :)

**Reply by vrothberg** [reply]
Yes, absolute! Quadlet also supports running K8s workloads (see https://docs.podman.io/en/latest/markdown/podman-systemd.unit.5.html#kube-units-kube).  That's also more elegant the pre-existing `podman-kube@` systemd template.

And the Quadlet files are easy to share, just like Compose or K8s YAML.

**Reply by mcdent** [reply]
sorry @vrothberg I am back again :) 

As I understand from what I have read, quadlet looks for a container unit file in `$HOME/.config/containers/systemd` (for rootless lets say).
Currently I have maybe 10 running containers and their unit files are in `$HOME/.config/systemd/user`

I initially started these containers from `podman run` commands, some of which were fairly length, volume mounts, ports etc.
(in retrospect I should have documented these). I was then able to use `podman generate systemd...` to create the unit files for them.

Is there any simple-ish way to move from my existing configuration to `quadlet` type unit files?

Am I right in thinking that the way to create a podman container based on quadlet, is to manually create the unit file in `$HOME/.config/containers/systemd` 
Then `systemctl --user daemon-reload` and `systemctl --user start myservice.service`

Or am I missing something?
Thanks again!

**Reply by vrothberg** [reply]
> Is there any simple-ish way to move from my existing configuration to quadlet type unit files?

There is no automated way, so there is some migration cost linked to it.  But I expect the 1-time migration to outweigh the continuous efforts of maintaining the generated units.

> Or am I missing something?

No, you are spot on.  You can drop a `name.container` in `$HOME/.config/containers/systemd`, reload the daemon and then you can start `name.service`.

**Comment by rhatdan** [comment]
https://github.com/containers/podman/pull/18891

**Reply by mcdent** [reply]
Wait, I was just searching for a way to specify the autoupdate feature in a Quadlet file :)
For the time do  I need to specify it as part of the `ExecStart=podman run --label io.containers.autoupdate=registry` ?

Does this mean in the next release it will be a native part of the quadlet unit file ?

Forgive my question if it is somewhat daft, I am still learning!

**Reply by vrothberg** [reply]
At the present you can set it as a `Label` in Quadlet but https://github.com/containers/podman/pull/18891 will make it more convenient.

Such conversations are a great source of inspiration.

**Reply by mcdent** [reply]
Thanks @vrothberg . I ran in to an issue where I had not specified the correct url/path to the image, so now my Quadlet file looks like:-
```
[Unit]
Description=Uptime Kuma
After=local-fs.target

[Container]
Image=registry.hub.docker.com/library/louislam/uptime-kuma:1
Volume=/home/mike/docker/uptime-kuma/run/data:/app/data
PublishPort=3001:3001
ContainerName=uptime-kuma
Label="io.containers.autoupdate=registry"
Exec=podman run

[Install]
# Start by default on boot
WantedBy=multi-user.target default.target
```

Hopefully that is too too far from being correct. The error I am seeing is:-

```
❯ systemctl --user status uptime-kuma.service
× uptime-kuma.service - Uptime Kuma
     Loaded: loaded (/home/mike/.config/containers/systemd/uptime-kuma.container; generated)
     Active: failed (Result: exit-code) since Wed 2023-06-14 18:49:08 BST; 8s ago
    Process: 329668 ExecStart=/usr/bin/podman run --name=uptime-kuma --cidfile=/run/user/1000/uptime-kuma.cid --replace --rm --log-driver passthrough --runtime /usr/bin/crun --cgroups=spli>
    Process: 329677 ExecStopPost=/usr/bin/podman rm -f -i --cidfile=/run/user/1000/uptime-kuma.cid (code=exited, status=0/SUCCESS)
    Process: 329686 ExecStopPost=rm -f /run/user/1000/uptime-kuma.cid (code=exited, status=0/SUCCESS)
   Main PID: 329668 (code=exited, status=125)
        CPU: 214ms

Jun 14 18:49:07 docker1 systemd[1408]: Starting Uptime Kuma...
Jun 14 18:49:07 docker1 uptime-kuma[329668]: Trying to pull registry.hub.docker.com/library/louislam/uptime-kuma:1...
Jun 14 18:49:08 docker1 uptime-kuma[329668]: Error: initializing source docker://registry.hub.docker.com/library/louislam/uptime-kuma:1: Requesting bearer token: invalid status code from r>
Jun 14 18:49:08 docker1 podman[329668]: 2023-06-14 18:49:07.753027474 +0100 BST m=+0.024107819 image pull  registry.hub.docker.com/library/louislam/uptime-kuma:1
Jun 14 18:49:08 docker1 systemd[1408]: uptime-kuma.service: Main process exited, code=exited, status=125/n/a
Jun 14 18:49:08 docker1 systemd[1408]: uptime-kuma.service: Failed with result 'exit-code'.
Jun 14 18:49:08 docker1 systemd[1408]: Failed to start Uptime Kuma.
```

Am I missing something obvious? Thanks!
